### PR TITLE
[DCP] Fixes the BC issue where the traversal doesn't support versions before 2.4

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134158

The original DCP doesn't flattening all the containers, which can cause issues, https://github.com/pytorch/pytorch/pull/125335 intends to solve the issue by flattening all the dictionaries.

Unfortunately, it breaks the checkpoints that are saved before 2.4. This
also shows some issues of the DCP:

1. DCP should record version in the metadata.
2. DCP should have a nice way to load old state_dict.
3. DCP should unflatten all containers (map, list) not just map.

This PR only addresses issue 2 to unblock users. Issue 1 and issue 3 need to be addressed in the future.

@pradeepfn Please let me know if this summary matches our discussion.

Fixes https://github.com/pytorch/pytorch/issues/133923

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn

### CODE REVIEW

**Review by wz337 — APPROVED** [review]
LGTM

File: torch/distributed/checkpoint/_version.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+
+from typing import Optional
+
+
+_act_like_version: Optional[str] = None
```
Comment (pradeepfn) [inline torch/distributed/checkpoint/_version.py:6]: nit: s/_act_like_version/derived_version

File: torch/distributed/checkpoint/_nested_dict.py [hunk 2]
```diff
@@ -40,7 +47,13 @@ def flat_copy(path: OBJ_PATH, value: STATE_DICT_ITEM) -> None:
         flattened[new_fqn] = value
         mappings[new_fqn] = path
 
-    traverse_state_dict(state_dict, flat_copy)
+    use_v_2_3 = (
```
Comment (pradeepfn) [inline torch/distributed/checkpoint/_nested_dict.py:50]: Shall we add a doc comment on the discrepancy we are solving. I see that you try to hit that by v_2_3, but a new devloper might not be able to understand the context.

File: torch/distributed/checkpoint/default_planner.py [hunk 3]
```diff
@@ -195,6 +197,20 @@ def set_up_planner(
...
+            missing_keys = load_keys - current_keys
+            if missing_keys:
+                _version._act_like_version = "2_3"
+                old_state_dict, old_mappings = flatten_state_dict(
+                    self.original_state_dict
+                )
+                old_keys = set(old_state_dict.keys())
+                if old_keys & missing_keys:
+                    self.state_dict, self.mappings = old_state_dict, old_mappings
+                _version._act_like_version = None
```
Comment (pradeepfn) [inline torch/distributed/checkpoint/default_planner.py:212]: I do not understand why we set the 

_version._act_like_version = 2.3 and then set the same to None at the end. 
Isn't that a noop ?

File: torch/distributed/checkpoint/default_planner.py [hunk 4]
```diff
@@ -195,6 +197,20 @@ def set_up_planner(
 
     def create_local_plan(self) -> LoadPlan:
         assert self.metadata is not None
+        if self.flatten_state_dict:
+            current_keys = set(self.state_dict.keys())
```
[the remark this answers is not in the document]
Comment (fegin) [reply torch/distributed/checkpoint/default_planner.py:201]: Please see the comment.

File: torch/distributed/checkpoint/default_planner.py (hunk 3 above)
Comment (fegin) [reply torch/distributed/checkpoint/default_planner.py:212]: Please see the comment.

File: torch/distributed/checkpoint/default_planner.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pradeepfn) [reply torch/distributed/checkpoint/default_planner.py:218]: Thanks for the detailed doc string!.

**Review by pradeepfn — APPROVED** [review]
How can we prevent BC breakages in future. Do you have insights on how to have a test suite that save with old code/load with new code for DCP code changes.

**Comment by fegin** [comment]
> How can we prevent BC breakages in future. Do you have insights on how to have a test suite that save with old code/load with new code for DCP code changes.

I think we should add some tests that generate checkpoints for 2.3. And this checkpoints shouldn't be changed (maybe use checksum to verify) regardless how the code base changes. We can then use this golden sample to verify if we can always load the golden sample back.

**Pull request closed without merging**